sway-core/src/semantic_analysis/storage_only_types.rs

```diff
@@ -109,6 +109,21 @@ fn expr_validate(expr: &TypedExpression) -> CompileResult<()> {
                 check!(expr_validate(f), (), warnings, errors);
             }
         }
+        TypedExpressionVariant::Reassignment(reassignment) => {
+            let TypedReassignment {
+                lhs_base_name, rhs, ..
+            } = &**reassignment;
+            check!(
+                check_type(rhs.return_type, lhs_base_name.span(), false),
+                (),
+                warnings,
+                errors,
+            );
+            check!(expr_validate(rhs), (), warnings, errors)
+        }
+        TypedExpressionVariant::StorageReassignment(_storage_reassignment) => {
+            // FIXME: there's nothing to check here?
```

Comment:
This is in `storage_only_types::expr_validate`. I'm not sure what this function is doing, but given we recurse into non-storage assignments for validation it seems like we should have to recurse into storage assignments to, no?

Reply:
This function is checking that there are no declarations of "storage-only" types. I don't know if syntactically we can have declarations inside here (but semantically, as per the AST definition, we can). So it wouldn't harm to recurse, but I'm not sure if it's necessary.

Reply:
> This function is checking that there are no declarations of "storage-only" types. I don't know if syntactically we can have declarations inside here (but semantically, as per the AST definition, we can). So it wouldn't harm to recurse, but I'm not sure if it's necessary.

Based on https://github.com/FuelLabs/sway/pull/2535#discussion_r949053844, I change my stance from "it wouldn't harm to recurse" to "we must recurse".